src/ripple/app/tx/impl/URIToken.cpp

```diff
@@ -0,0 +1,819 @@
+            // check if the buyer has the right trustline with an adequate
+            // balance
+
+            STAmount availableFunds{accountFunds(
+                ctx.view, acc, purchaseAmount, fhZERO_IF_FROZEN, ctx.j)};
+
+            if (purchaseAmount > availableFunds)
+                return tecINSUFFICIENT_FUNDS;
+
+            return tesSUCCESS;
+        }
+
+        case ttURITOKEN_CANCEL_SELL_OFFER: {
+            if (acc != *owner)
+                return tecNO_PERMISSION;
+
+            return tesSUCCESS;
+        }
+
+        case ttURITOKEN_CREATE_SELL_OFFER: {
+            if (acc != *owner)
+                return tecNO_PERMISSION;
+
+            STAmount txAmount = ctx.tx.getFieldAmount(sfAmount);
+            if (!txAmount.native())
+            {
+                AccountID const iouIssuer = txAmount.getIssuer();
+                if (!ctx.view.exists(keylet::account(iouIssuer)))
+                    return tecNO_ISSUER;
+            }
+            return tesSUCCESS;
+        }
+
+        default: {
+            JLOG(ctx.j.warn()) << "URIToken txid=" << ctx.tx.getTransactionID()
+                               << " preclaim with tt = " << tt << "\n";
+            return tecINTERNAL;
+        }
+    }
+}
+
+TER
+URIToken::doApply()
+{
+    auto j = ctx_.app.journal("View");
+
+    auto const sle = view().peek(keylet::account(account_));
+    if (!sle)
+        return tefINTERNAL;
+
+    uint16_t tt = ctx_.tx.getFieldU16(sfTransactionType);
+
+    if (tt == ttURITOKEN_MINT || tt == ttURITOKEN_BUY)
+    {
+        STAmount const reserve{
+            view().fees().accountReserve(sle->getFieldU32(sfOwnerCount) + 1)};
+
+        if (mPriorBalance - ctx_.tx.getFieldAmount(sfFee).xrp() < reserve)
+            return tecINSUFFICIENT_RESERVE;
+    }
+
+    uint32_t flags = ctx_.tx.getFlags();
+
+    std::shared_ptr<SLE> sleU;
+    std::optional<AccountID> issuer;
+    std::optional<AccountID> owner;
+    std::optional<STAmount> saleAmount;
+    std::optional<AccountID> dest;
+    std::optional<Keylet> kl;
+    std::shared_ptr<SLE> sleOwner;
+
+    if (tt != ttURITOKEN_MINT)
+    {
+        kl = Keylet{ltURI_TOKEN, ctx_.tx.getFieldH256(sfURITokenID)};
+        sleU = view().peek(*kl);
+
+        if (!sleU)
+            return tecNO_ENTRY;
+
+        if (sleU->getFieldU16(sfLedgerEntryType) != ltURI_TOKEN)
+            return tecNO_ENTRY;
+
+        owner = (*sleU)[sfOwner];
+        issuer = (*sleU)[sfIssuer];
+        saleAmount = (*sleU)[~sfAmount];
+        dest = (*sleU)[~sfDestination];
+
+        if (*owner == account_)
+            sleOwner = sle;
+        else
+            sleOwner = view().peek(keylet::account(*owner));
+
+        if (!sleOwner)
+        {
+            JLOG(j.warn()) << "Malformed transaction: owner of URIToken is not "
+                              "in the ledger.";
+            return tecNO_ENTRY;
+        }
+    }
+
+    switch (tt)
+    {
+        case ttURITOKEN_MINT: {
```

Comment:
what do you think about making each of these switch bodies their own private methods, for readability?